### PR TITLE
androidenv: update repo.json with a new strategy to expire

### PR DESCRIPTION
**Pull request by hadilq**
###### Description of changes

This PR is solving the problem with different approaches of nixpkgs vs Android SDK to deliver the packages. Android SDK fetches the newest packages, via XML files, and let the user install them, but it'll not remove the packages that are not available in those XML files, so those old packages could co-exist with the new ones, but in our approach in nixpkgs, the user doesn't have access to the old packages, that are not available in the XML files.

With this PR I targeted that problem, also to show this approach in action, **I updated packages** using this new methods.

Therefore, in these changes the `mkrepo.rb` file will receive the old `repo.json` file via `STDIN`, then expire packages that are removed from the XML files more than two years ago, then merge the old JSON with the new ones, the ones that are fetched from the XML files.

###### Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [X] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [X] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [X] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by hadilq** [comment]
@numinit Hey Morgan, if you have time please review this one. Thanks.

**Comment by numinit** [comment]
Interesting approach, I've bumped into this before. Does the Android repo XML have a way to tell us when the packages expire? And, just to be clear, the two year limit is for packages that have _disappeared_ altogether, correct?

**Comment by hadilq** [comment]
@numinit sorry I respond late!

> Interesting approach, I've bumped into this before. Does the Android repo XML have a way to tell us when the packages expire?
- As far as I know, there's no way to know when the links in the google servers expire, that's why I put a random number, two years.  Some of the packages are very old, but I tried to solve the problem related to the ones that disappear very fast, like the `emulator` ones.

> And, just to be clear, the two year limit is for packages that have _disappeared_ altogether, correct?
- Yes, they _disappeared_ from the XML files.

Just for your information, I'm not pushing this strategy to be merged, I spent time on it because it's a real problem to solve, and I enjoyed coding it.

**Comment by hadilq** [comment]
Hey @numinit, I noticed I've forgot to ask about you and other's opinions :D Sorry! Do you even agree the problem exist? If yes, do you think it's the solution, or maybe there are other/better solutions? If yes, I really want to know about them, so I could update this PR, or close it and create another one. Is there any group that I could notify them to gather their opinion about it? Thanks.

**Comment by numinit** [comment]
I have seen this problem and think this is a good fix actually. Sorry it took me so long to get back to you, my work (where we use androidenv, actually) picked up in the last couple weeks.

**Review by numinit — APPROVED** [review]

**Comment by hadilq** [comment]
@numinit No worries. If I [asked for reviewers](https://github.com/NixOS/nixpkgs/issues/230371), because I understood the situation, so there's no complain. 

Awesome. Thank you for your review.

File: pkgs/development/mobile/androidenv/mkrepo.sh [hunk 1]
```diff
@@ -6,14 +6,16 @@ set -e
 pushd "$(dirname "$0")" &>/dev/null || exit 1
 
 echo "Writing repo.json" >&2
-ruby mkrepo.rb \
+REPO_JSON="$(cat ./repo.json)"
+echo $REPO_JSON | ruby mkrepo.rb \
```
Comment (SuperSandro2000) [inline pkgs/development/mobile/androidenv/mkrepo.sh:10]:
```suggestion
cat ./repo.json | ruby mkrepo.rb \
```

File: pkgs/development/mobile/androidenv/mkrepo.sh [hunk 2]
```diff
@@ -6,14 +6,16 @@ set -e
...
     --packages ./xml/repository2-1.xml \
     --images ./xml/android-sys-img2-1.xml \
     --images ./xml/android-tv-sys-img2-1.xml \
     --images ./xml/android-wear-cn-sys-img2-1.xml \
     --images ./xml/android-wear-sys-img2-1.xml \
     --images ./xml/google_apis-sys-img2-1.xml \
     --images ./xml/google_apis_playstore-sys-img2-1.xml \
-    --addons ./xml/addon2-1.xml > repo.json
+    --addons ./xml/addon2-1.xml \
+         > repo.json
```
Comment (SuperSandro2000) [inline pkgs/development/mobile/androidenv/mkrepo.sh:19]:
```suggestion
    --addons ./xml/addon2-1.xml \
      | sponge repo.json
```

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 1 above)
Comment (hadilq) [reply pkgs/development/mobile/androidenv/mkrepo.sh:10]: In fact, I considered `cat` in the first place, and it took me a while to notice `ruby` cannot handle the STDIN from `cat`! The worst thing is that it doesn't raise an exception/error, the STDIN is just empty! My mental model is that `ruby` is slower than `cat`, so there's a back pressure, which non of them handled it, but when you load the entire data into RAM, there's no pressure and ruby can handle it. Let me know if there's any other way around.

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 2 above)
Comment (hadilq) [reply pkgs/development/mobile/androidenv/mkrepo.sh:19]: I got 
```
./mkrepo.sh: line 18: sponge: command not found
```
and apparently `sponge` is not available in `nixpkgs`! Let me know how can I resolve it.

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/development/mobile/androidenv/mkrepo.sh:10]: You could just load the file from within the ruby script.

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 2 above)
Comment (SuperSandro2000) [reply pkgs/development/mobile/androidenv/mkrepo.sh:19]: sponge is part of moreutils

```
 ▶ nix-locate bin/sponge
moreutils.out                                    18,400 x /nix/store/5zh3jp0q5w108111ppcwjxrpbxsy3wsl-moreutils-0.67/bin/sponge
```

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 1 above)
Comment (hadilq) [reply pkgs/development/mobile/androidenv/mkrepo.sh:10]: I have two reasons to not doing it:
 - I'm not the creator of `mkrepo.sh`, or `mkrepo.rb` files, but as far as I understand from their implementation, there's an abstraction is going on. The `mkrepo.rb` is supposed to be agnostic about the where the input is coming from and where the output is going to.
 - My second reason is that I already tried that, and I tried it again after your suggestion, but without success! This is my code
 ```ruby
  input_json = ""
  if File.file? 'repo.json'
    $stderr.write('yes')
  end
  File.foreach('repo.json') { |line|
      input_json += line
  }
  $stderr.write(input_json)
 ```
Its output is just
```console
yes%
```
so `input_json` is empty! Even that `File.foreach` is supposed to be okay with very long files. I have no clue why it's not working! Ruby is not the language that I work with everyday, so maybe I missed something! @numinit I noticed you are much more experienced than me with Ruby. Do you have any idea?

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 2 above)
Comment (hadilq) [reply pkgs/development/mobile/androidenv/mkrepo.sh]: Thanks, fixed in https://github.com/NixOS/nixpkgs/pull/227666/commits/de2679819e0ab01b9488b151efe4a2d433f1fa25 It also solved the problem with empty stdin above!

File: pkgs/development/mobile/androidenv/mkrepo.sh (hunk 1 above)
Comment (hadilq) [reply pkgs/development/mobile/androidenv/mkrepo.sh]: After I noticed what's `sponge`, that @SuperSandro2000 suggested below, I could fix it in https://github.com/NixOS/nixpkgs/pull/227666/commits/de2679819e0ab01b9488b151efe4a2d433f1fa25. So in the end, it was not the Ruby problem, but bash is to blame! Thanks @SuperSandro2000.

**Review by numinit — APPROVED** [review]

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Pull request merged**